### PR TITLE
Allow main package to be at non-root path of app

### PR DESCRIPTION
**Pull request by SemanticallyNull**
It was not possible to build an application if it had a main package that was not in the root of the workspace. This PR creates an environment variable which will allow users to set a path.

Introduces BP_APP_MAIN_PATH environmental variable
- defaults to "." if BP_APP_MAIN_PATH not set

With @iainsproat

### CODE REVIEW

**Comment by sclevine** [comment]
Sorry for the delay! We're still working on Tracker integration for the new buildpacks.

I think we'd like this to be configurable via a `buildpack.yml` file instead of an environment variable. Maybe something like this:

```yaml
gomod:
  paths: ["some/cmd/here", "another/cmd/overhere"] 
```

**Comment by scothis** [comment]
It's common to have multiple targets within a single repo that should each be compiled into separate images. The target package should be configurable per build.

**Comment by sclevine** [comment]
That makes sense. I think we should also support multiple targets in the same image as well (e.g., for extra binaries for DB migrations, debug tools, etc.). The v2 version of the buildpack support multiple targets via `GO_INSTALL_PACKAGE_SPEC`.

We're trying to capture all of the build configuration in buildpack.yml and only introduce environment variables where parameters could reasonably vary across builds. I agree that this is one of those cases -- I should have considered that in my previous comment.

How about we support a default set of targets in buildpack.yml:
```yaml
go-mod:
  targets: ["some/cmd/here", "another/cmd/overhere"] 
```
As well as an environment variable to override that set:
```
BP_GO_MOD_TARGETS=some/cmd/different:another/cmd/overhere
```

**Comment by scothis** [comment]
@sclevine sounds good.

If there are multiple targets, how do you set the entry point?

**Comment by sclevine** [comment]
`CMD` is set to run the binary built from the first target.

**Comment by SemanticallyNull** [comment]
@sclevine the update that we just pushed should fix the functionality to what you described in a previous comment.

File: mod/mod.go [hunk 1]
```diff
@@ -154,7 +175,47 @@ func (c Contributor) setStartCommand() error {
...
+		targets := strings.Split(buildTarget, ":")
+		return targets, nil
+	}
+
+	configPath := filepath.Join(c.appRoot, "buildpack.yml")
+	config := Config{}
+	if _, err := os.Stat(configPath); err == nil {
+		yamlFile, err := ioutil.ReadFile(configPath)
+		if err != nil {
+			return []string{}, err
```
Comment (sclevine) [inline mod/mod.go]: Small stylistic comment: use `return nil, err` instead

File: mod/mod_test.go [hunk 2]
```diff
@@ -85,6 +90,46 @@ func testGoMod(t *testing.T, when spec.G, it spec.S) {
...
+					})
+				})
+
+				when("buildpack.yml config file is present", func() {
+					it("runs `go install`, gets app name and contributes the start command", func() {
+						Expect(ioutil.WriteFile(filepath.Join(appRoot, "buildpack.yml"),
+							[]byte(`---
+go-mod:
+  targets: ["./path/to/my"]`),
+							os.FileMode(0600))).To(Succeed())
```
Comment (sclevine) [inline mod/mod_test.go]: Use `0666` and rely on OS umask for non-owner adjustments.

**Comment by sclevine** [comment]
This is looking great!

We discussed the format of buildpack.yml today and decided that language-specific tooling should be nested under a language key. Happy for whoever merges to adjust this, but it should be either:
```
go:
  mod:
    targets: ["./path/to/my"]
```
or:
```
go:
  targets: ["./path/to/my"]
```
depending on if the targets concept is general enough to apply to all ways of building go apps (e.g., dep). I think it is, so I support the second option.

**Comment by SemanticallyNull** [comment]
We've made those changes, we used the latter as the way to do things. 

We're currently investigating an issue where the environment variable doesn't actually work, despite the tests passing. Once we've figured it out we'll update the PR and add a test to make sure this doesn't happen again.

**Pull request merged**